packages/cli/src/util/logs.ts

```diff
@@ -30,11 +38,217 @@ export function displayBuildLogs(
+    signal: abortController.signal,
+    retry: {
+      retries: 3,
+      onRetry: err => {
+        log(`Runtime logs error: ${err.message}`);
+        if (err instanceof Error && err.name === 'AbortError') {
+          return;
+        }
+        throw err;
+      },
+    },
+  });
+  // handle the event stream and make the promise get rejected
+  // if errors occur so we can retry
+  return new Promise<number>((resolve, reject) => {
+    const stream = response.body.pipe(parse ? jsonlines.parse() : split());
+    let finished = false;
+    let errored = false;
+
+    function finish(err?: unknown) {
+      if (finished) return;
+      clearTimeout(timeout);
+      stopSpinner();
+      finished = true;
+      if (err) {
+        reject(err);
+      } else {
+        resolve(abortController.signal.aborted ? 1 : 0);
+      }
+    }
+
+    const handleData = (data: RuntimeLog | string) => {
+      let log: RuntimeLog = parse ? data : JSON.parse(data as string);
+      if (isRuntimeLimitDelimiter(log)) {
+        abortController.abort();
+        warn(`${chalk.bold(log.message)}\n`);
+        return;
+      }
+      // eslint-disable-next-line no-console -- we intent to write unparsed logs to stdout so JQ could read them
+      parse ? prettyPrintLogline(log, print) : console.log(data);
+      spinner(runtimeLogSpinnerMessage);
+    };
+
+    const handleError = (err: Error) => {
+      if (finished || errored) return;
+      if (err.name === 'AbortError') {
+        finish();
+        return;
+      }
+      stream.destroy();
+      errored = true;
+      debug(`Runtime logs stream error: ${err.message ?? err}`);
+
+      setTimeout(() => {
+        if (abortController.signal.aborted) return;
+        // retry without maximum amount nor clear past logs etc
+        displayRuntimeLogs(client, options, abortController).then(
+          resolve,
+          reject
+        );
+      }, 2000);
+    };
+
+    stream.on('end', finish);
+    stream.on('data', handleData);
+    stream.on('error', handleError);
+    response.body.on('error', handleError);
+  });
+}
+
+function printBuildLog(log: any, print: Printer) {
+  if (!log.created) return; // keepalive
+
+  let data: string;
+
+  data = (log.text || '')
+    .replace(/\n$/, '')
+    .replace(/^\n/, '')
+    // eslint-disable-next-line no-control-regex
+    .replace(/\x1b\[1000D/g, '')
+    .replace(/\x1b\[0K/g, '')
+    .replace(/\x1b\[1A/g, '');
+  if (/warning/i.test(data)) {
+    data = chalk.yellow(data);
+  } else if (log.type === 'stderr') {
+    data = chalk.red(data);
+  }
+
+  const date = new Date(log.created).toISOString();
+
+  data.split('\n').forEach(line => {
+    print(`${chalk.dim(date)}  ${line.replace('[now-builder-debug] ', '')}\n`);
+  });
+}
+
+function isRuntimeLimitDelimiter(log: RuntimeLog) {
+  return (
+    log.rowId === '' && log.level === 'error' && log.source === 'delimiter'
```

Comment:
**question**: wouldn't `delimiter` source be sufficient check here

Reply:
I don't think so. We may have other delimiters, it doesn't harm being more specific in this case, to match [our API endpoint convention](https://github.com/vercel/api/pull/29618/files#diff-ee9519d82b888d4d4fd7e3b14ef4cd0b6f0d84d71fd81399cfd9417040de7ad1R58-R71)